File: src/main/java/nl/wernerkroneman/Drawy/ModelEditor/Knowledge.java
package nl.wernerkroneman.Drawy.ModelEditor;

import nl.wernerkroneman.Drawy.Modelling.PrimitiveModel;
import nl.wernerkroneman.Drawy.Modelling.Model;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.Map;
import java.util.TreeMap;

public class Knowledge {

    // @inv For each entry: model name matches key
    Map<String, Model> knownObjects = new TreeMap<String, Model>(String.CASE_INSENSITIVE_ORDER);

    public static Knowledge knowledgeWithPrimitives() {
        Knowledge knowledge = new Knowledge();

        knowledge.remember(new PrimitiveModel(PrimitiveModel.ShapeType.CUBE, "Cube"));
        knowledge.remember(new PrimitiveModel(PrimitiveModel.ShapeType.CYLINDER, "Cylinder"));
        knowledge.remember(new PrimitiveModel(PrimitiveModel.ShapeType.SPHERE, "Sphere"));

        return knowledge;
    }

    public Model getObject(String name) {
        return knownObjects.get(name);
    }

    public void remember(Model model) {
        knownObjects.put(model.getName(), model);
    }

    public int getNumberOfObjects() {
        return knownObjects.size();
    }

    public boolean isKnownObject(Model model) {
        return knownObjects.values().stream().anyMatch(kO -> model.equals(kO));
    }

}

File: src/main/java/nl/wernerkroneman/Drawy/ModelEditor/Knowledge.kt
/*
 * Copyright (c) 2017 Werner Kroneman
 *
 * This file is part of DrawyMcDrawface.
 *
 * DrawyMcDrawface is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DrawyMcDrawface is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with DrawyMcDrawface.  If not, see <http://www.gnu.org/licenses/>.
 */

package nl.wernerkroneman.Drawy.ModelEditor

import nl.wernerkroneman.Drawy.Modelling.ModelSpecification
import nl.wernerkroneman.Drawy.Modelling.PrimitiveModelSpecification
import nl.wernerkroneman.Drawy.Modelling.PrimitiveModelSpecificationBase
import java.util.*

class Knowledge {

    // @inv For each entry: model name matches key
    var knownObjects: MutableMap<String, ModelSpecification> = TreeMap(String.CASE_INSENSITIVE_ORDER)

    fun getObject(name: String): ModelSpecification {
        return knownObjects[name] ?:
                throw NoSuchElementException("I don't know any model by the name of $name.")
    }

    fun remember(model: ModelSpecification) {
        knownObjects.put(model.name, model)
    }

    val numberOfObjects: Int
        get() = knownObjects.size

    fun isKnownObject(model: ModelSpecification): Boolean {
        return knownObjects.values.any({ it == model })
    }

    companion object {

        fun knowledgeWithPrimitives(): Knowledge {
            val knowledge = Knowledge()

            knowledge.remember(
                    PrimitiveModelSpecificationBase(name = "Cube", shape = PrimitiveModelSpecification.ShapeType.CUBE))
            knowledge.remember(
                    PrimitiveModelSpecificationBase(name = "Cylinder", shape = PrimitiveModelSpecification.ShapeType.CYLINDER))
            knowledge.remember(
                    PrimitiveModelSpecificationBase(name = "Sphere", shape = PrimitiveModelSpecification.ShapeType.SPHERE))
            knowledge.remember(
                    PrimitiveModelSpecificationBase(name = "Cone", shape = PrimitiveModelSpecification.ShapeType.CONE))

            return knowledge
        }
    }

}